src/lib/plugins/GraphQLPlugin/server/HttpClient.ts

```diff
@@ -0,0 +1,159 @@
+import 'isomorphic-fetch'
+import { forEach, toPairs } from 'lodash'
+import * as qs from 'querystring'
+import { InjectionContext, InjectionClient } from '../../../core/InjectionClient'
+
+export class HttpError implements Error {
+  readonly status: number
+
+  constructor(status: number) {
+    this.status = status
+  }
+
+  get message() {
+    return `Http ${this.status}`
+  }
+
+  get name() {
+    return 'HTTPError'
+  }
+}
+
+export type HttpMethod = 'GET' | 'PUT' | 'PATCH' | 'POST' | 'DELETE'
+
+export interface FetchUrl {
+  baseUrl: string
+  path: string
+  query?: {}
+}
+
+export interface RequestOpts {
+  url: string | FetchUrl
+  body?: any
+  headers?: Record<string, string>
+  responseType?: ResponseType
+  bodyType?: BodyType
+}
+
+export interface ResponseType {
+  headers: Record<string, string>
+  convert: (x: Response) => Promise<any> | undefined
+}
+
+export interface BodyType {
+  headers: Record<string, string>
+  convert: (x: any) => any
+}
+
+export const BodyType = {
+  FormData: {
+    convert: (body: Record<string, string>) => {
+      const form = new FormData()
+      forEach(body, (value, key) => {
+        form.append(key, value)
+      })
+      return form
+    },
+    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+  },
+  Json: {
+    convert: JSON.stringify,
+    headers: { 'Content-Type': 'application/json' }
+  },
+  Empty: {
+    convert: () => undefined,
+    headers: {}
+  }
+}
+
+export const ResponseType = {
+  Json: {
+    convert: (x: Response) => x.json(),
+    headers: { Accept: 'application/json' }
+  },
+  Empty: {
+    convert: () => undefined,
+    headers: {}
+  }
+}
+
+export interface HttpClientProps {
+  fetchFn?: (input: RequestInfo, init?: RequestInit) => Promise<Response>
+  context: InjectionContext
+}
+
+export class HttpClient extends InjectionClient {
+  private fetch: (input: RequestInfo, init?: RequestInit) => Promise<Response>
+
+  constructor({ fetchFn = fetch, context }: HttpClientProps) {
+    super(context)
+    this.fetch = fetchFn
+  }
+
+  async request(method: HttpMethod, opts: RequestOpts) {
+    const { headers, body, bodyType = BodyType.Empty, responseType = ResponseType.Empty } = opts
+    const res = await this.fetch(formatUrl(opts.url), {
+      method,
+      body: bodyType.convert(body),
+      headers: toPairs({
+        ...(headers || {}),
```

Comment:
So in this case, you're not able to override the framework's choice of headers.

If I wanted to have an application/json+customtype

(which is accepted by browsers as application.json) I wouldn't be able to override what the framework had decided should be a application.json header

Can we put 'custom' headers at the bottom, so their value takes precedence

Reply:
Yeah, explicitly specified headers should probable take prescience 👍

Reply:
Although you are able to override the framework's choice of headers by providing a different body/response type.

Reply:
Yes I saw that afterwards - but even so, for convenience I'd still say users should be able to override it

Reply:
Yeah, that's the behaviour I'd expect